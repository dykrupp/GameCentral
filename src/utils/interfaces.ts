import { FluidObject } from 'gatsby-image';
import { Product } from 'use-shopping-cart';

export interface ProductInfo extends Product {
  fluidObject: FluidObject;
  type: string;
  isTopPick: boolean | null;
}
