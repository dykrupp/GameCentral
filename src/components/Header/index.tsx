import { Link } from 'gatsby';
import PropTypes from 'prop-types';
import React, { useState } from 'react';
import { useShoppingCart } from 'use-shopping-cart';
import styled from 'styled-components';

interface HeaderPropTypes {
  siteTitle?: string;
}

const Button = styled.button`
  font-size: 13px;
  text-align: center;
  color: #fff;
  outline: none;
  padding: 12px;
  box-shadow: 2px 5px 10px rgba(0, 0, 0, 0.1);
  background-color: rgb(255, 178, 56);
  border-radius: 6px;
  letter-spacing: 1.5px;
  width: 150px;
`;

const CartInfo = styled.p`
  margin-bottom: 0;
  font-weight: bold;
  color: white;
`;

const Header: React.FC<HeaderPropTypes> = ({ siteTitle }) => {
  const [loading, setLoading] = useState(false);

  const {
    formattedTotalPrice,
    redirectToCheckout,
    cartCount,
    clearCart,
  } = useShoppingCart();

  return (
    <header
      style={{
        background: `rebeccapurple`,
        marginBottom: `1.45rem`,
      }}
    >
      <div
        style={{
          margin: `0 auto`,
          padding: `1.45rem 1.0875rem`,
          display: 'flex',
          justifyContent: 'space-between',
          alignItems: 'center',
        }}
      >
        <h1 style={{ margin: 0 }}>
          <Link
            to="/"
            style={{
              color: `white`,
              textDecoration: `none`,
            }}
          >
            {siteTitle}
          </Link>
        </h1>

        <CartInfo>Number of Items: {cartCount}</CartInfo>
        <CartInfo>Total: {formattedTotalPrice}</CartInfo>
        <Button
          disabled={loading}
          onClick={() => {
            setLoading(true);
            redirectToCheckout();
          }}
        >
          {loading ? 'Loading...' : 'Checkout'}
        </Button>
        <Button onClick={clearCart}>Clear Cart</Button>
      </div>
    </header>
  );
};

Header.propTypes = {
  siteTitle: PropTypes.string,
};

Header.defaultProps = {
  siteTitle: ``,
};

export default Header;
