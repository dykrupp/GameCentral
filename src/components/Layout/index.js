import React from 'react';
import { loadStripe } from '@stripe/stripe-js';
import { CartProvider } from 'use-shopping-cart';
import { ThemeProvider, createMuiTheme } from '@material-ui/core/styles';
import { CssBaseline } from '@material-ui/core';
import InnerLayout from './InnerLayout';
import ScrollToTop from 'react-scroll-to-top';
import './global.css';

const stripePromise = loadStripe(
  process.env.GATSBY_STRIPE_PUBLISHABLE_KEY
    ? process.env.GATSBY_STRIPE_PUBLISHABLE_KEY
    : ''
);

const palette = {
  primary: { main: '#673AB7' },
  secondary: { main: '#03a9f4' },
};

const typography = {
  fontFamily: 'Arial',
};

export default ({ children, location }) => (
  <CartProvider
    mode="client-only"
    stripe={stripePromise}
    successUrl={`${location.origin}/purchase-confirmation/`}
    cancelUrl={`${location.origin}/`}
    currency="USD"
    allowedCountries={['US', 'CA']}
    billingAddressCollection={true}
  >
    <ThemeProvider theme={createMuiTheme({ palette, typography })}>
      <CssBaseline />
        <InnerLayout>
          <ScrollToTop smooth color="#673AB7"/>
          {children}
        </InnerLayout>
    </ThemeProvider>
  </CartProvider>
);
