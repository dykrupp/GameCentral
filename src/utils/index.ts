export const formatPrice = (amount: number, currency: string): string => {
  const price = parseFloat((amount / 100).toFixed(2));
  const numberFormat = new Intl.NumberFormat(['en-US'], {
    style: 'currency',
    currency: currency,
    currencyDisplay: 'symbol',
  });
  return numberFormat.format(price);
};
