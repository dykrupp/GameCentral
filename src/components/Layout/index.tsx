import React, { ReactNode } from 'react';
import PropTypes from 'prop-types';
import { useStaticQuery, graphql } from 'gatsby';
import Header from './Header';
import styled from 'styled-components';
import './layout.css';

interface LayoutProps {
  children: ReactNode;
}

const LayoutContainer = styled.div`
  height: 100%;
  display: flex;
  flex-direction: column;
`;

const MainContainer = styled.main`
  flex-grow: 1;
`;

const Footer = styled.footer`
  text-align: center;
`;

const Layout: React.FC<LayoutProps> = ({ children }) => {
  const data = useStaticQuery(graphql`
    query SiteTitleQuery {
      site {
        siteMetadata {
          title
        }
      }
    }
  `);

  return (
    <LayoutContainer>
      <Header siteTitle={data.site.siteMetadata.title} />
      <MainContainer>{children}</MainContainer>
      <Footer>
        Â© {new Date().getFullYear()}, Built with
        {` `}
        <a href="https://www.gatsbyjs.org">Gatsby</a>
      </Footer>
    </LayoutContainer>
  );
};

Layout.propTypes = {
  children: PropTypes.node.isRequired,
};

export default Layout;
