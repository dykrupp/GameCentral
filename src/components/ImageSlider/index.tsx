import React, { useState } from 'react';
import AwesomeSlider from 'react-awesome-slider';
import 'react-awesome-slider/dist/styles.css';
import 'react-awesome-slider/dist/custom-animations/cube-animation.css';
import sliderStyles from './styles.module.css';
import styled from 'styled-components';
import { usePickedProductInfo } from '../../utils/hooks/usePickedProductInfo';
import Skeleton from '@material-ui/lab/Skeleton';
import { ProductInfo } from '../../utils/interfaces';
import PropTypes from 'prop-types';

export const sliderHeight = '350px';

const ImageSliderContainer = styled.div`
  height: ${sliderHeight};
`;

const LoadingSkeleton = styled(Skeleton)`
  height: ${sliderHeight};
  transform: none;
`;

interface ImageSliderProps {
  setCurrentProduct: (product: ProductInfo) => void;
}

export const ImageSlider: React.FC<ImageSliderProps> = ({
  setCurrentProduct,
}) => {
  const [isLoading, setIsLoading] = useState(true);
  const pickedProductInfo = usePickedProductInfo();

  return (
    <>
      {isLoading && <LoadingSkeleton animation="wave" />}
      <ImageSliderContainer style={{ display: isLoading ? 'none' : 'block' }}>
        <AwesomeSlider
          infinite={true}
          animation="cubeAnimation"
          bullets={true}
          cssModule={sliderStyles}
          onFirstMount={(e) => {
            setCurrentProduct(pickedProductInfo[e.currentIndex]);
            setIsLoading(false);
          }}
          onTransitionRequest={(e) =>
            setCurrentProduct(pickedProductInfo[e.nextIndex])
          }
        >
          {pickedProductInfo.map((productInfo) => (
            <div key={productInfo.sku} data-src={productInfo.fluidObject.src} />
          ))}
        </AwesomeSlider>
      </ImageSliderContainer>
    </>
  );
};

ImageSlider.propTypes = {
  setCurrentProduct: PropTypes.func.isRequired,
};
