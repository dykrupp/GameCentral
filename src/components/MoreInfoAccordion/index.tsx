import React, { useState, useEffect } from 'react';
import { Theme, createStyles, makeStyles } from '@material-ui/core/styles';
import Accordion from '@material-ui/core/Accordion';
import AccordionSummary from '@material-ui/core/AccordionSummary';
import AccordionDetails from '@material-ui/core/AccordionDetails';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import { ProductInfo } from '../../utils/interfaces';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { ProductType } from '../../utils/constants';
import Skeleton from '@material-ui/lab/Skeleton';

const AccordionProductDetails = styled(AccordionDetails)`
  text-align: center;
  flex-direction: column;
`;

const TitleText = styled(Typography)`
  margin-bottom: 10px;
  margin-top: -15px;
`;

const MoreInfoText = styled(Typography)`
  margin-bottom: 10px;
`;

const MoreInfoSkeleton = styled(Skeleton)`
  height: 75px;
`;

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    accordion: {
      width: '100%',
      backgroundColor: theme.palette.primary.main,
      color: 'white',
    },
    heading: {
      fontSize: theme.typography.pxToRem(15),
      fontWeight: theme.typography.fontWeightRegular,
    },
    summary: {
      '& .MuiAccordionSummary-content': {
        justifyContent: 'center',
        marginLeft: '30px',
      },
    },
  })
);

const toApiPlatformFormat = (type: string): string => {
  switch (type) {
    case ProductType.PS4:
      return 'playstation-4';
    case ProductType.Nintendo:
      return 'switch';
    case ProductType.PC:
    case ProductType.VR:
      return 'pc';
    case ProductType.Xbox:
      return 'XONE';
    default:
      return 'UNKNOWN';
  }
};

interface MetacriticInfo {
  availableOn: [string];
  description: string;
  developer: string;
  genre: [string];
  image: string;
  publisher: [string];
  rating: string;
  releaseDate: string;
  score: number;
  title: string;
}

interface MoreInfoAccordionProps {
  productInfo: ProductInfo | null;
  setIsPlaying?: (value: boolean) => void;
}

export const MoreInfoAccordion: React.FC<MoreInfoAccordionProps> = ({
  productInfo,
  setIsPlaying,
}) => {
  const classes = useStyles();
  const [isLoading, setIsLoading] = useState(true);
  const [metaCriticInfo, setMetaCriticInfo] = useState<null | MetacriticInfo>(
    null
  );

  useEffect(() => {
    if (!productInfo || !process.env.GATSBY_METACRITIC_API_KEY) return;

    //needed to prevent errors being thrown if components is unmounted while in the process of 'fetching'
    let didCancel = false;

    setIsLoading(true);

    fetch(
      `https://chicken-coop.p.rapidapi.com/games/${
        productInfo.name
      }?platform=${toApiPlatformFormat(productInfo.type)}`,
      {
        method: 'GET',
        headers: {
          'x-rapidapi-host': 'chicken-coop.p.rapidapi.com',
          'x-rapidapi-key': process.env.GATSBY_METACRITIC_API_KEY,
        },
      }
    )
      .then((response) => response.json())
      .then((data) => {
        if (didCancel) return;
        else if (data && data.result && data.result !== 'No result') setMetaCriticInfo(data.result as MetacriticInfo);
        else setMetaCriticInfo(null);
      })
      .catch((err) => {
        console.log(err);
      })
      .finally(() => {
        if (!didCancel) setIsLoading(false);
      });

    return () => {
      didCancel = true;
    };
  }, [productInfo, process.env.GATSBY_METACRITIC_API_KEY]);

  if (!productInfo) return null;
  return (
    <Accordion
      className={classes.accordion}
      onClick={() => setIsPlaying && setIsPlaying(false)}
    >
      <AccordionSummary
        style={{ display: 'flex'}}
        className={classes.summary}
        expandIcon={<ExpandMoreIcon style={{ color: 'white' }} />}
      >
        <Typography className={classes.heading}>MORE INFORMATION</Typography>
      </AccordionSummary>
      {isLoading ? (
        <SkeltonDetails />
      ) : metaCriticInfo !== null ? (
        <MetacriticDetails
          description={productInfo.description ? productInfo.description : ''}
          metacriticInfo={metaCriticInfo}
          platform={productInfo.type}
        />
      ) : (
        <ProductDetails productInfo={productInfo} />
      )}
    </Accordion>
  );
};

MoreInfoAccordion.propTypes = {
  productInfo: PropTypes.any,
  setIsPlaying: PropTypes.func,
};

const SkeltonDetails = () => (
  <AccordionProductDetails>
    <MoreInfoSkeleton />
  </AccordionProductDetails>
);

const ProductDetails: React.FC<MoreInfoAccordionProps> = ({ productInfo }) => {
  if (!productInfo) return null;
  return (
    <AccordionProductDetails>
      <TitleText>{`Title: ${productInfo.name}`}</TitleText>
      <Typography>{`Description: `}</Typography>
      <Typography>{productInfo.description}</Typography>
      <MoreInfoText>
        {productInfo ? `Platform: ${productInfo.type}` : ''}
      </MoreInfoText>
    </AccordionProductDetails>
  );
};

ProductDetails.propTypes = {
  productInfo: PropTypes.any,
};

interface MetacriticDetailsProps {
  metacriticInfo: MetacriticInfo;
  description: string;
  platform: string;
}

const MetacriticDetails: React.FC<MetacriticDetailsProps> = ({
  metacriticInfo,
  description,
  platform,
}) => (
  <AccordionProductDetails>
    <TitleText>{`Title: ${metacriticInfo.title}`}</TitleText>
    <Typography>{`Description: `}</Typography>
    <MoreInfoText>{description}</MoreInfoText>
    <MoreInfoText>{`Platform: ${platform}`}</MoreInfoText>
    <MoreInfoText>{`Developer: ${metacriticInfo.developer}`}</MoreInfoText>
    <MoreInfoText>{`Genres: ${metacriticInfo.genre.join(', ')}`}</MoreInfoText>
    <MoreInfoText>{`Release Date: ${metacriticInfo.releaseDate}`}</MoreInfoText>
    <MoreInfoText>{`Metacritic Score: ${metacriticInfo.score}`}</MoreInfoText>
    <MoreInfoText>{`Rating: ${metacriticInfo.rating}`}</MoreInfoText>
  </AccordionProductDetails>
);

MetacriticDetails.propTypes = {
  metacriticInfo: PropTypes.any.isRequired,
  description: PropTypes.string.isRequired,
  platform: PropTypes.string.isRequired,
};
