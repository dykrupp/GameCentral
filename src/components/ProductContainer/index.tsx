import React from 'react';
import { graphql, useStaticQuery } from 'gatsby';
import styled from 'styled-components';
import ProductItem from './ProductItem';
import { FluidObject } from 'gatsby-image';
import { Product } from 'use-shopping-cart';
import { ProductType } from '../../utils/types';
import PropTypes from 'prop-types';

const Container = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: space-between;
  padding: 1rem 0 1rem 0;
`;

interface PriceNode {
  product: {
    name: string;
    description: string;
    id: string;
    images: string[];
    metadata: {
      type: string;
    };
  };
  id: string;
  unit_amount: number;
  currency: string;
  active: boolean;
}

interface imageLocalFile {
  parent: {
    id: string;
  };
  childImageSharp: {
    fluid: FluidObject;
  };
}

interface ImageNode {
  localFiles: imageLocalFile[];
}

interface ProductContainerProps {
  type: ProductType;
}

const ProductContainer: React.FC<ProductContainerProps> = ({ type }) => {
  const { prices, images } = useStaticQuery(
    graphql`
      query ProductInfo {
        prices: allStripePrice {
          nodes {
            product {
              name
              description
              id
              images
              metadata {
                type
              }
            }
            id
            unit_amount
            currency
            active
          }
        }
        images: allStripeProduct {
          nodes {
            localFiles {
              parent {
                id
              }
              childImageSharp {
                fluid(maxWidth: 150) {
                  ...GatsbyImageSharpFluid
                }
              }
            }
          }
        }
      }
    `
  );

  const imageNodes = images.nodes as ImageNode[];
  const priceNodes = (prices.nodes as PriceNode[]).filter(
    (priceNode) => priceNode.active
  );

  return (
    <Container>
      {priceNodes.map((priceNode, index) => {
        if (type !== priceNode.product.metadata.type) return null;

        const newProduct: Product = {
          name: priceNode.product.name,
          sku: priceNode.id,
          price: priceNode.unit_amount,
          currency: priceNode.currency,
          description: priceNode.product.description,
          image: priceNode.product.images[0],
        };

        const currentFluidImage = imageNodes.find(
          (x) => x.localFiles[0].parent.id === priceNode.product.id
        );

        if (!currentFluidImage) {
          console.log(`Not showing ${newProduct.name}, due to missing image`);
          return null;
        }

        return (
          <ProductItem
            key={index}
            product={newProduct}
            fluidImage={currentFluidImage.localFiles[0].childImageSharp.fluid}
          />
        );
      })}
    </Container>
  );
};

ProductContainer.propTypes = {
  type: PropTypes.string.isRequired,
};

export default ProductContainer;
