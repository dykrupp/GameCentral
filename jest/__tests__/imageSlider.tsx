import React from 'react';
import { render } from '@testing-library/react';
import { PureImageSlider } from '../../src/components/ImageSlider';
import { ProductInfo } from '../../src/utils/interfaces';

const newProduct: ProductInfo = {
  name: 'World of Warcraft',
  type: 'PC',
  sku: 'abc123ghsdsd',
  price: 20.99,
  currency: 'USD',
  isTopPick: false,
  fluidObject: {
    aspectRatio: 1.7936,
    src: 'test-src',
    srcSet: '/sample-src-set',
    sizes: '(max-width: 450px) 100vw, 450px',
  },
};

describe('ImageSlider', () => {
  it('renders', () => {
    const setIsPlaying = jest.fn();
    const setCurrentProduct = jest.fn();

    const { asFragment } = render(
      <PureImageSlider
        isPlaying={false}
        setIsPlaying={setIsPlaying}
        setCurrentProduct={setCurrentProduct}
        productInfo={[newProduct]}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('autoplays on initialization', () => {
    const setIsPlaying = jest.fn();
    const setCurrentProduct = jest.fn();

    const { getByTestId } = render(
      <PureImageSlider
        isPlaying={true}
        setIsPlaying={setIsPlaying}
        setCurrentProduct={setCurrentProduct}
        productInfo={[newProduct]}
      />
    );

    const imageSliderEle = getByTestId('image-slider-container').children[0];

    const wrapperEle = Array.from(imageSliderEle.children).filter(
      (x) => x.className === 'awssld__wrapper'
    )[0];

    const divChildren = Array.from(
      Array.from(wrapperEle.children).filter(
        (x) => x.className === 'awssld__container'
      )[0].children
    );

    const activeContentDiv = Array.from(divChildren).filter((x) =>
      x.classList.contains('awssld--active')
    )[0].children[0];

    const activeTimer = Array.from(activeContentDiv.children).filter((x) =>
      x.classList.contains('awssld__timer')
    )[0];

    expect(activeTimer).toBeDefined();
  });
});
