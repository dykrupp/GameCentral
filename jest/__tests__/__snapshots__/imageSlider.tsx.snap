// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSlider renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm fdgoAF"
  >
    <div
      class="sc-AxirZ iWxEsR"
      data-testid="image-slider-container"
      style="display: block;"
    >
      <div
        class="awssld--organic-arrows awssld"
      >
        <div
          class="awssld__wrapper"
        >
          <div
            class="awssld__container"
          >
            <div
              class="awssld__box awssld--active"
            >
              <div
                class="awssld__content"
              >
                <img
                  src="test-src"
                />
              </div>
            </div>
            <div
              class="awssld__box"
            />
          </div>
          <div
            class="awssld__controls awssld__controls--hidden"
          >
            <button
              aria-label="next"
              class="awssld__next"
            >
              <span
                class="awssld__controls__arrow-right"
              />
            </button>
            <button
              aria-label="previous"
              class="awssld__prev"
            >
              <span
                class="awssld__controls__arrow-left"
              />
            </button>
          </div>
        </div>
        <nav
          class="awssld__bullets"
        >
          <button
            class="awssld__bullets--active"
            data-index="0"
          >
            0
          </button>
        </nav>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
